### ISSUE TITLE
DevX upgrading from 2.0.0 to latest main is problematic

### ISSUE THREAD

**Issue by abailly-iohk** [input-output-hk/quickcheck-dynamic#30]
I have tried to upgrade Hydra's `Model` to latest main version (aef136ede729a25762aecf95979951c07add817a) and it's somewhat annoying: 
* I needed to add a `VarContext` to `arbitraryAction` which I don't use: Why should I pay for something I don't care about? 
* I then had a cryptic error telling me there were no `Arbitrary` instances on the `Action WorldState a` type
* This led me to realise I needed to add the `makeActionVariables` TH splice which in itself is a bit annoying: Having to rely on TH leads to longer compilation time
* Adding the missing splice requires also importing `HasVariables(..)` which is not at all obvious: In the examples, we import the full module which is not a good practice (it makes it hard to track where things are going from just by looking at the file)
* More importantly, it does not compile because there's a lot of missing `Generic` instances for various types which are deeply nested in the `Action` constructors, sometimes even types which are not our own
* Trying to manually add an instance for `HasVariables` also leads to this `Generic` problem

All in all, these changes introduce a lot of ergonomic issues which should be addressed before we can consider a new release.

**Comment by abailly-iohk** [comment]
Perhaps a satisfying solution that would have the added advantage of not breaking existing code would be to have a child typeclass for `StateModel` that cares about variables tracking, leaving the plain `StateModel` untouched?

**Comment by MaximilianAlgehed** [comment]
> I needed to add a VarContext to arbitraryAction which I don't use: Why should I pay for something I don't care about?
You do care about this if you want to generate symbolic things that are in scope and well-typed. This sounds like a problem you're having while updating - not a huge problem as we don't have many users?

> This led me to realise I needed to add the makeActionVariables TH splice which in itself is a bit annoying: Having to rely on TH leads to longer compilation time
Yes, this is annoying I agree and it would be nice to find a generic-programming solution to this but the problem is that `Generic` doesn't like GADTs.

> Trying to manually add an instance for HasVariables also leads to this Generic problem.
Can you give some more details about this?

**Comment by abailly-iohk** [comment]
>  This sounds like a problem you're having while updating - not a huge problem as we don't have many users?

Sure, but we want more, don't we? I am playing the role of the guinea pig here, the average developer who wants to use q-d and could be baffled by too much technical complexity, on top of the conceptual complexity of understanding how to properly use a Model-Based Testing approach. One of the reasons I have never used q-s-m is that I found its machinery overly complicated.

> Can you give some more details about this?

I found the solution: Adding a `getAllVariables = const mempty` method removes the constraint on `Generic`

**Comment by MaximilianAlgehed** [comment]
> I found the solution: Adding a getAllVariables = const mempty method removes the constraint on Generic

What do you mean, where do you add this? In the instance declarations?

How much of this complexity do you think good documentation can overcome? The problem is that the complexity we are introducing here is *already there*, just that right now it's a footgun. If you use symbolic variables you need to ensure the well-scopedness and well-typedness by manually maintaining the variable context in the model. This clutters the model and is all-around horrible. We are trying to get rid of this problem by doing the tracking for you.

I genuinely don't think this is complexity that we can avoid so the question to my mind is what interface would make it most paletable.

**Comment by abailly-iohk** [comment]
> What do you mean, where do you add this? In the instance declarations?

Yes. This removes the dependency on the default implementation requiring `Generic`. My `Model` now compiles but I am having troubles in the `ModelSpec` now.

> How much of this complexity do you think good documentation can overcome?

I think it's definitely possible to overcome, and perhaps someone coming to this afresh would not be as annoyed as me because I have to upgrade existing code.
I feel the usage of symbolic variables is the least well explained part in the existing docs, perhaps also because I did not have a need for them so I did not spend the time to understand really how they work and write a good doc for that. And you're the experts there so not too well placed to put yourself in the shoes of an average dev :) 

> what interface would make it most paletable.

I don't know how much of this is possible, but what do you think of my suggestion about having 2 layers of abstraction? A "Simple" one with only basic `StateModel` w/o variables, and another one say `SymbolicStateModel` taking into account variables definition and tracking? 

Perhaps I simply need to familiarise myself more with the use of `Var` in order to be able to explain them in layman's terms.

**Comment by MaximilianAlgehed** [comment]
> I don't know how much of this is possible, but what do you think of my suggestion about having 2 layers of abstraction?

I can see this being a nice way to do things as it would simulatenously remove the footgun and reduce needless complexity when you're not using symbolics.

I think it's a good idea to involve @UlfNorell in this conversation.

**Comment by abailly-iohk** [comment]
This afternoon I hope to find some time to try to port some tests in Consensus from q-s-m to q-d, and perhaps there will be a need for `Var`iables there so that I can make sense of their use :)

**Comment by MaximilianAlgehed** [comment]
You need them when you need some information that is only available at runtime to generate actions. In the registry example for instance you get `ThreadId`s at runtime and you want to generate action sequences like:
```
do
  tid <- action $ Spawn
  action $ Register "a" tid
  action $ Register "b" tid
```
But `tid` is not available as a `ThreadId` when you generate the action sequence, it only becomes available at runtime so you need to represent it symbolically.

Now, you might think you don't need this phase separation between generation and running - why not generate the program "on the fly" - once you've run `Spawn` you do in fact have a `ThreadId` so you could just use that! However, this approach is fundamentally flawed - you can't shrink the counterexamples you get from this testing technique! And QuickCheck without shrinking is just crap.

**Comment by abailly-iohk** [comment]
I know what they are useful for @MaximilianAlgehed 😄 Just saying I did not have the need _yet_, but then that's exactly the thing we want to explain in the docs.

**Comment by abailly-iohk** [comment]
Another problem I am running into:
```
test/Hydra/ModelSpec.hs:137:27: error:
    • No instance for (Generic (VerificationKey Hydra.Crypto.HydraKey))
        arising from a use of ‘forAllQ’
      There are instances for similar types:
        instance Generic
                   cardano-crypto-wrapper-1.3.0:Cardano.Crypto.Signing.VerificationKey.VerificationKey
          -- Defined in ‘cardano-crypto-wrapper-1.3.0:Cardano.Crypto.Signing.VerificationKey’
    • In a stmt of a 'do' block:
        (party, payment) <- forAllQ (nonConflictingTx st)
      In the expression:
        do (party, payment) <- forAllQ (nonConflictingTx st)
           void $ action $ Model.NewTx party payment
           eventually (ObserveConfirmedTx payment)
      In a case alternative:
          st@WorldState {hydraState = Open {}}
            -> do (party, payment) <- forAllQ (nonConflictingTx st)
                  void $ action $ Model.NewTx party payment
                  eventually (ObserveConfirmedTx payment)
    |
137 |       (party, payment) <- forAllQ (nonConflictingTx st)
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

There's a `HasVariables` constraint on `Quantifiable` which means every value we generate as part of a DL formula needs to be an instance of `HasVariables`

**Comment by MaximilianAlgehed** [comment]
Indeed. That's another snag that you hit upon as a consequence of doing symbolic variables right - it turns out you have to be very careful about how you shrink DL formulae. I suppose there is some super clever reason that `VerificationKey` isn't an instance of `Generic`?

**Comment by abailly-iohk** [comment]
Well, I have no idea, that's something which comes from the underlying crypto library. I agree most of the time one wants `Generic` as part of standard `deriving` but 🤷

**Comment by MaximilianAlgehed** [comment]
@UlfNorell do you want to weigh in on this?

**Comment by abailly-iohk** [comment]
FWIW I managed to do the upgrade so the question is more rhetorical than practical. Perhaps having some more documentation would be just as fine.

**Comment by MaximilianAlgehed** [comment]
FWIW 3.1.1 is supposed to have fixed some of the confusing error messages here.
